sky/data/csync/redirect-fuse.c

```diff
@@ -0,0 +1,542 @@
+	} else if (read) {
+        result = full_read_path;
+        free(full_write_path);
+	} else if (file_exists(full_write_path)) {
+        result = full_write_path;
+        free(full_read_path);
+    } else {
+        result = full_read_path;
+        free(full_write_path);
+    } 
+
+    return result;
+}
+
+static void *xmp_init(struct fuse_conn_info *conn,
+		      struct fuse_config *cfg)
+{
+	(void) conn;
+	cfg->use_ino = 1;
+
+	/* Pick up changes from lower filesystem right away. This is
+	   also necessary for better hardlink support. When the kernel
+	   calls the unlink() handler, it does not know the inode of
+	   the to-be-removed entry and can therefore not invalidate
+	   the cache of the associated inode - resulting in an
+	   incorrect st_nlink value being reported for any remaining
+	   hardlinks to this inode. */
+	cfg->entry_timeout = 0;
+	cfg->attr_timeout = 0;
+	cfg->negative_timeout = 0;
+
+	return NULL;
+}
+
+static int xmp_getattr(const char *path, struct stat *stbuf,
+		       struct fuse_file_info *fi)
+{
+
+	(void) fi;
+    char* translated_path = full_path(path, 0, 0);
+    int res = lstat(translated_path, stbuf);
+    free(translated_path);
+	if (res == -1) {
+		return -errno;
+	}
+	return 0;
+}
+
+static int xmp_access(const char *path, int mask)
+{
+    char* translated_path = full_path(path, 0, 0);
+    int res = access(translated_path, mask);
+	free(translated_path);
+	if (res == -1) {
+		return -errno;
+	}
+	return 0;
+}
+
+
+static int xmp_readdir(const char *path, void *buf, fuse_fill_dir_t filler, off_t offset, struct fuse_file_info *fi, enum fuse_readdir_flags flags) {
+    DIR *dp;
+    struct dirent *de;
+
+    char* read_path = full_path(path, 0, 1);
+    char* write_path = full_path(path, 1, 0);
+
+    // Clear hash table
+	SimpleSet set;
+	set_init(&set);
+    // Process read_path
+    dp = opendir(read_path);
+    if (dp != NULL) {
+        while ((de = readdir(dp)) != NULL) {
+			if (set_contains(&set, de->d_name) == SET_FALSE) {
+                struct stat st;
+                memset(&st, 0, sizeof(st));
+                st.st_ino = de->d_ino;
+                st.st_mode = de->d_type << 12;
+                if (filler(buf, de->d_name, &st, 0, fill_dir_plus)) {
+                    break;
+				}
+				set_add(&set, de->d_name);
+            }
+        }
+        closedir(dp);
+    }
+
+    // Process write_path
+    dp = opendir(write_path);
+    if (dp != NULL) {
+        while ((de = readdir(dp)) != NULL) {
+			if (set_contains(&set, de->d_name) == SET_FALSE) {
+                struct stat st;
+                memset(&st, 0, sizeof(st));
+                st.st_ino = de->d_ino;
+                st.st_mode = de->d_type << 12;
+                if (filler(buf, de->d_name, &st, 0, fill_dir_plus)) {
+                    break;
+				}
+				set_add(&set, de->d_name);
+            }
+        }
+        closedir(dp);
+    }
+
+    free(read_path);
+    free(write_path);
+    set_destroy(&set);
```

Comment:
why do we need this set? is it possible that under one dir there are two files with same name?

Reply:
It is possible to have two files with same name in each directories.

Reply:
sry can you elaborate a little bit?

Reply:
So `xmp_readdir` file operation is called when `ls` is ran at the mountpoint. And we design this in a way that `ls` displays all the files/directories from both read directory(mounted by cloud storage) and write directory(the write cache). And if there's a file with identical names in both directories, we only want to display it once.

Reply:
Why not only display the read dir? IIUC it should include every file in the write dir?

Reply:
Oh is this for the case when there are some new write files in the write dir that have not been synced to the cloud?